File: python/bc-components/src/bc_components/encrypted_key/_pbkdf2_params.py
"""PBKDF2 key derivation parameters."""

from __future__ import annotations

from bc_crypto import pbkdf2_hmac_sha256, pbkdf2_hmac_sha512
from bc_tags import CBOR

from .._salt import Salt
from ..symmetric._encrypted_message import EncryptedMessage
from ..symmetric._symmetric_key import SymmetricKey
from ._hash_type import HashType
from ._hkdf_params import SALT_LEN
from ._key_derivation_method import KeyDerivationMethod


class PBKDF2Params:
    """PBKDF2 key derivation parameters.

    CDDL::

        PBKDF2Params = [1, Salt, iterations: uint, HashType]
    """

    INDEX: int = KeyDerivationMethod.PBKDF2.value

    __slots__ = ("_salt", "_iterations", "_hash_type")

    def __init__(self, salt: Salt, iterations: int, hash_type: HashType) -> None:
        self._salt = salt
        self._iterations = iterations
        self._hash_type = hash_type

    # --- Construction ---

    @staticmethod
    def generate() -> PBKDF2Params:
        """Create new PBKDF2 params with defaults (100000 iterations, SHA256)."""
        return PBKDF2Params(
            Salt.generate_with_len(SALT_LEN),
            100_000,
            HashType.SHA256,
        )

    @staticmethod
    def new_opt(salt: Salt, iterations: int, hash_type: HashType) -> PBKDF2Params:
        """Create new PBKDF2 params with specific parameters."""
        return PBKDF2Params(salt, iterations, hash_type)

    # --- Properties ---

    @property
    def salt(self) -> Salt:
        return self._salt

    @property
    def iterations(self) -> int:
        return self._iterations

    @property
    def hash_type(self) -> HashType:
        return self._hash_type

    # --- Key derivation ---

    def _derive_key(self, secret: bytes | bytearray) -> SymmetricKey:
        """Derive a symmetric key using PBKDF2."""
        secret_bytes = bytes(secret)
        if self._hash_type == HashType.SHA256:
            derived = pbkdf2_hmac_sha256(
                secret_bytes, self._salt.data, self._iterations, 32
            )
        else:
            derived = pbkdf2_hmac_sha512(
                secret_bytes, self._salt.data, self._iterations, 32
            )
        return SymmetricKey.from_data(derived)

    def lock(
        self,
        content_key: SymmetricKey,
        secret: bytes | bytearray,
    ) -> EncryptedMessage:
        """Derive a key from *secret* via PBKDF2 and encrypt *content_key*."""
        derived_key = self._derive_key(secret)
        encoded_method = self.to_cbor().to_cbor_data()
        return derived_key.encrypt(content_key.data, aad=encoded_method)

    def unlock(
        self,
        encrypted_message: EncryptedMessage,
        secret: bytes | bytearray,
    ) -> SymmetricKey:
        """Derive a key from *secret* via PBKDF2 and decrypt the content key."""
        derived_key = self._derive_key(secret)
        plaintext = derived_key.decrypt(encrypted_message)
        return SymmetricKey.from_data(plaintext)

    # --- Display ---

    def __str__(self) -> str:
        return f"PBKDF2({self._hash_type})"

    def __repr__(self) -> str:
        return (
            f"PBKDF2Params(salt={self._salt!r}, "
            f"iterations={self._iterations}, "
            f"hash_type={self._hash_type!r})"
        )

    def __eq__(self, other: object) -> bool:
        if isinstance(other, PBKDF2Params):
            return (
                self._salt == other._salt
                and self._iterations == other._iterations
                and self._hash_type == other._hash_type
            )
        return NotImplemented

    def __hash__(self) -> int:
        return hash((self._salt, self._iterations, self._hash_type))

    # --- CBOR ---

    def to_cbor(self) -> CBOR:
        """Encode as CBOR array: [INDEX, salt, iterations, hash_type]."""
        return CBOR.from_array([
            CBOR.from_int(self.INDEX),
            self._salt.untagged_cbor(),
            CBOR.from_int(self._iterations),
            self._hash_type.to_cbor(),
        ])

    @staticmethod
    def from_cbor(cbor: CBOR) -> PBKDF2Params:
        """Decode from CBOR array: [INDEX, salt, iterations, hash_type]."""
        a = cbor.try_array()
        if len(a) != 4:
            raise ValueError("Invalid PBKDF2Params: expected 4 elements")
        salt = Salt.from_untagged_cbor(a[1])
        iterations = a[2].try_int()
        hash_type = HashType.from_cbor(a[3])
        return PBKDF2Params(salt, iterations, hash_type)

File: typescript/bc-components/src/encrypted-key/pbkdf2-params.ts
import {
    pbkdf2HmacSha256,
    pbkdf2HmacSha512,
} from '@bc/crypto';
import { type Cbor, cbor as toCbor } from '@bc/dcbor';

import { Salt } from '../salt.js';
import { SymmetricKey } from '../symmetric/symmetric-key.js';
import type { EncryptedMessage } from '../symmetric/encrypted-message.js';
import { HashType, hashTypeFromCbor, hashTypeToCbor, hashTypeToString } from './hash-type.js';
import type { KeyDerivation } from './key-derivation.js';
import { KeyDerivationMethod } from './key-derivation-method.js';
import { SALT_LEN } from './hkdf-params.js';

export class PBKDF2Params implements KeyDerivation {
    readonly #salt: Salt;
    readonly #iterations: number;
    readonly #hashType: HashType;

    constructor(salt: Salt, iterations: number, hashType: HashType) {
        this.#salt = salt;
        this.#iterations = iterations;
        this.#hashType = hashType;
    }

    static new(): PBKDF2Params {
        return PBKDF2Params.newOpt(
            Salt.newWithLen(SALT_LEN),
            100_000,
            HashType.SHA256,
        );
    }

    static newOpt(
        salt: Salt,
        iterations: number,
        hashType: HashType,
    ): PBKDF2Params {
        return new PBKDF2Params(salt, iterations, hashType);
    }

    salt(): Salt {
        return this.#salt;
    }

    iterations(): number {
        return this.#iterations;
    }

    hashType(): HashType {
        return this.#hashType;
    }

    index(): number {
        return KeyDerivationMethod.PBKDF2 as number;
    }

    lock(contentKey: SymmetricKey, secret: Uint8Array): EncryptedMessage {
        const derivedKey = SymmetricKey.fromData(
            this.#hashType === HashType.SHA256
                ? pbkdf2HmacSha256(
                    secret,
                    this.#salt.asBytes(),
                    this.#iterations,
                    32,
                )
                : pbkdf2HmacSha512(
                    secret,
                    this.#salt.asBytes(),
                    this.#iterations,
                    32,
                ),
        );
        const encodedMethod = this.toCbor().toData();
        return derivedKey.encrypt(contentKey.data, encodedMethod);
    }

    unlock(encryptedMessage: EncryptedMessage, secret: Uint8Array): SymmetricKey {
        const derivedKey = SymmetricKey.fromData(
            this.#hashType === HashType.SHA256
                ? pbkdf2HmacSha256(
                    secret,
                    this.#salt.asBytes(),
                    this.#iterations,
                    32,
                )
                : pbkdf2HmacSha512(
                    secret,
                    this.#salt.asBytes(),
                    this.#iterations,
                    32,
                ),
        );
        return SymmetricKey.fromData(derivedKey.decrypt(encryptedMessage));
    }

    toCbor(): Cbor {
        return toCbor([
            KeyDerivationMethod.PBKDF2,
            this.#salt.taggedCbor(),
            this.#iterations,
            hashTypeToCbor(this.#hashType),
        ]);
    }

    static fromCbor(cbor: Cbor): PBKDF2Params {
        const a = cbor.toArray();
        if (a.length !== 4) {
            throw new Error('Invalid PBKDF2Params');
        }
        const salt = Salt.fromCbor(a[1]!);
        const iterations = Number(a[2]!.toInteger());
        const hashType = hashTypeFromCbor(a[3]!);
        return new PBKDF2Params(salt, iterations, hashType);
    }

    toString(): string {
        return `PBKDF2(${hashTypeToString(this.#hashType)})`;
    }
}
